In [ ]:
https://www.kaggle.com/datasets/yasserh/advertising-sales-dataset/data

In [ ]:
import pandas as pd
import numpy as np

df = pd.read_csv('Advertising Budget and Sales.csv')
print(df.head())
print(df.info())
print(df.describe())

   Unnamed: 0  TV Ad Budget ($)  Radio Ad Budget ($)  Newspaper Ad Budget ($)  \
0           1             230.1                 37.8                     69.2   
1           2              44.5                 39.3                     45.1   
2           3              17.2                 45.9                     69.3   
3           4             151.5                 41.3                     58.5   
4           5             180.8                 10.8                     58.4   

   Sales ($)  
0       22.1  
1       10.4  
2        9.3  
3       18.5  
4       12.9  
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 200 entries, 0 to 199
Data columns (total 5 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   Unnamed: 0               200 non-null    int64  
 1   TV Ad Budget ($)         200 non-null    float64
 2   Radio Ad Budget ($)      200 non-null    float64
 3   Newspaper Ad Budget ($)  200 non-null    flo

In [ ]:
if 'Unamed: 0' in df.columns:
  df = df.drop(columns=['Unnamed: 0'])
X = df[['TV Ad Budget ($)', 'Radio Ad Budget ($)', 'Newspaper Ad Budget ($)']]
y = df['Sales ($)']

from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)
print("taille du jeu d'entraînement (X_train) :", X_train.shape)
print("taille du jeu de test (X_test) :", X_test.shape)

taille du jeu d'entraînement (X_train) : (160, 3)
taille du jeu de test (X_test) : (40, 3)


In [ ]:
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error, r2_score

model = LinearRegression()
model.fit(X_train, y_train)

y_pred = model.predict(X_test)

mse = mean_squared_error(y_test, y_pred)
rmse = np.sqrt(mse)
r2 = r2_score(y_test, y_pred)

print("performance de la régression linéaire")
print(f"RMSE (erreur moyenne) : {rmse:.4f}")
print(f"Score R2 (coefficienet de détermination) : {r2:.4f}")

coefficients = pd.DataFrame (model.coef_, X.columns, columns=['Coefficient'])
print("\nCoefficients du modèle :")
print(coefficients)
print(f"Intercept (biais de base) : {model.intercept_:.4f}")

performance de la régression linéaire
RMSE (erreur moyenne) : 1.7816
Score R2 (coefficienet de détermination) : 0.8994

Coefficients du modèle :
                         Coefficient
TV Ad Budget ($)            0.044730
Radio Ad Budget ($)         0.189195
Newspaper Ad Budget ($)     0.002761
Intercept (biais de base) : 2.9791


Score $R^2$ (0.899) : Le modèle explique 90% de la variance des ventes. Cela signifie que les budgets publicitaires sélectionnés capturent presque toute l'information nécessaire pour prédire le niveau des ventes.
RMSE (1.7816) : En moyenne, les prédictions du modèle s'écartent des ventes réelles d'environ 1,8 unité Interprétation des coefficients:
Radio (0.1892) : C'est le canal qui offre le meilleur rendement par dollar investi.
TV (0.0447) : Un impact positif et solide.
Newspaper (0.0028) : Son coefficient est très proche de 0, ce qui indique que la publicité dans la presse écrite a un impact quasi nul ou non significatif sur les ventes dans ce modèle.

In [ ]:
from sklearn.ensemble import RandomForestRegressor
from sklearn.metrics import mean_squared_error, r2_score

rf_model = RandomForestRegressor(n_estimators=100, random_state=42)
rf_model.fit(X_train, y_train)

y_pred_rf = rf_model.predict(X_test)
mse_rf = mean_squared_error(y_test, y_pred_rf)
rmse_rf = np.sqrt(mse_rf)
r2_rf = r2_score(y_test, y_pred)

print("Performance du random forest")
print(f"RMSE (erreur moy) : {rmse_rf:.4f}")
print(f"Score R2 (coeff de détermination) : {r2_rf:.4f}")

Performance du random forest
RMSE (erreur moy) : 0.7686
Score R2 (coeff de détermination) : 0.8994


In [ ]:
import xgboost as xgb

xgb_model = xgb.XGBRegressor(n_estimators=100, learning_rate=0.1, random_state=42)

xgb_model.fit(X_train, y_train)
y_pred_xgb = xgb_model.predict(X_test)

mse_xgb = mean_squared_error(y_test, y_pred_xgb)
rmse_xgb = np.sqrt(mse_xgb)
r2_xgb = r2_score(y_test, y_pred_xgb)

print("Performance de XGBOOST")
print(f"RMSE (erreur moy) : {rmse_xgb:.4f}")
print("fScore R2 'coeff de détermination) : {r2_xgb:.4f}")
print(r2_xgb)

Performance de XGBOOST
RMSE (erreur moy) : 0.8828
fScore R2 'coeff de détermination) : {r2_xgb:.4f}
0.9753082787568474


XGBoost s'impose comme le meilleur modèle pour prédire les ventes avec la meilleure précision.

In [ ]:
import joblib

joblib.dump(xgb_model, 'xgboost_advertising_model.pkl')
print("modèle XGBoost sauvegardé sous 'xgboost_advertising_model.pkl'")

modèle XGBoost sauvegardé sous 'xgboost_advertising_model.pkl'
